File: packages/core/src/marginalia_core/ports/llm.py
"""LLM-facing ports for rewrite and summarization."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Protocol

from marginalia_core.ports.capabilities import ProviderCapabilities


@dataclass(frozen=True, slots=True)
class RewriteInstruction:
    """Input passed to a rewrite provider."""

    document_title: str
    section_title: str
    source_anchor: str
    section_text: str
    note_texts: tuple[str, ...]


@dataclass(frozen=True, slots=True)
class RewriteOutput:
    """Structured rewrite output."""

    provider_name: str
    rewritten_text: str
    strategy: str
    note_count: int


@dataclass(frozen=True, slots=True)
class SummaryInstruction:
    """Input passed to a summarization provider."""

    topic: str
    matched_document_ids: tuple[str, ...]
    context_excerpt: str = ""


@dataclass(frozen=True, slots=True)
class SummaryOutput:
    """Structured summary output."""

    provider_name: str
    summary_text: str
    highlights: tuple[str, ...] = ()
    confidence: float = 1.0


class RewriteGenerator(Protocol):
    """Generate a section rewrite from source text and notes."""

    def describe_capabilities(self) -> ProviderCapabilities:
        """Describe rewrite provider behavior and constraints."""
        ...

    def rewrite_section(self, instruction: RewriteInstruction) -> RewriteOutput:
        """Return rewritten text for a section."""
        ...


class TopicSummarizer(Protocol):
    """Summarize a topic across the local corpus."""

    def describe_capabilities(self) -> ProviderCapabilities:
        """Describe summarizer behavior and constraints."""
        ...

    def summarize_topic(self, instruction: SummaryInstruction) -> SummaryOutput:
        """Return a topic summary."""
        ...

File: crates/marginalia-core/src/ports/llm.rs
use crate::ports::capabilities::ProviderCapabilities;

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct RewriteInstruction {
    pub document_title: String,
    pub section_title: String,
    pub source_anchor: String,
    pub section_text: String,
    pub note_texts: Vec<String>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct RewriteOutput {
    pub provider_name: String,
    pub rewritten_text: String,
    pub strategy: String,
    pub note_count: usize,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct SummaryInstruction {
    pub topic: String,
    pub matched_document_ids: Vec<String>,
    pub context_excerpt: String,
}

#[derive(Debug, Clone, PartialEq)]
pub struct SummaryOutput {
    pub provider_name: String,
    pub summary_text: String,
    pub highlights: Vec<String>,
    pub confidence: f64,
}

pub trait RewriteGenerator {
    fn describe_capabilities(&self) -> ProviderCapabilities;
    fn rewrite_section(&self, instruction: RewriteInstruction) -> RewriteOutput;
}

pub trait TopicSummarizer {
    fn describe_capabilities(&self) -> ProviderCapabilities;
    fn summarize_topic(&self, instruction: SummaryInstruction) -> SummaryOutput;
}
